sharding/xorValidatorShuffler.go

```diff
@@ -96,134 +103,181 @@ func (rxs *randXORShuffler) UpdateNodeLists(args ArgsUpdateNodes) ResUpdateNodes
 	}
 
 	// TODO: remove the conditional and the else branch when reinitialization of node in new shard is implemented
+	var validatorsDistributor ValidatorsDistributor = nil
 	if rxs.shuffleBetweenShards {
-		return shuffleNodesInterShards(eligibleAfterReshard, waitingAfterReshard, leavingNodes, args.NewNodes, args.Rand)
+		validatorsDistributor = &CrossShardValidatorDistributor{}
+	} else {
+		validatorsDistributor = &IntraShardValidatorDistributor{}
 	}
 
-	return shuffleNodesIntraShards(eligibleAfterReshard, waitingAfterReshard, leavingNodes, args.NewNodes, args.Rand)
+	return shuffleNodes(
+		eligibleAfterReshard,
+		waitingAfterReshard,
+		args.UnstakeLeaving,
+		args.AdditionalLeaving,
+		args.NewNodes,
+		args.Rand,
+		validatorsDistributor,
+	)
+}
+
+func removeDupplicates(unstake []Validator, additionalLeaving []Validator) {
+	for _, validator := range unstake {
```

Comment:
please use a different naming for validator as it shadows struct name

Reply:
fixed